File: Сортировка_шелла.py
def shell_sort(arr):
    n = len(arr)
    gap = n // 2  # Начальный шаг
    
    while gap > 0:
        for i in range(gap, n):
            temp = arr[i]  # Сохраняем текущий элемент
            j = i
            # Сдвигаем элементы, пока не найдем правильную позицию
            while j >= gap and arr[j - gap] > temp:
                arr[j] = arr[j - gap]
                j -= gap
            arr[j] = temp  # Вставляем сохраненный элемент
        gap //= 2  # Уменьшаем шаг

if __name__ == "__main__":
    arr = [12, 34, 54, 2, 3]
    print("Исходный массив:", arr)
    shell_sort(arr)
    print("Отсортированный массив:", arr)

File: Сортировка_шелла.java
import java.util.Arrays;

public class ShellSort {
    public static void shellSort(int[] arr) {
        int n = arr.length;
        
        // Начинаем с большого шага и уменьшаем его
        for (int gap = n / 2; gap > 0; gap /= 2) {
            // Применяем сортировку вставками для этого шага
            for (int i = gap; i < n; i++) {
                int temp = arr[i]; // Сохраняем текущий элемент
                int j;
                
                // Сдвигаем элементы, пока не найдем правильную позицию
                for (j = i; j >= gap && arr[j - gap] > temp; j -= gap) {
                    arr[j] = arr[j - gap];
                }
                arr[j] = temp; // Вставляем сохраненный элемент
            }
        }
    }
    
    public static void main(String[] args) {
        int[] arr = {12, 34, 54, 2, 3};
        System.out.println("Исходный массив: " + Arrays.toString(arr));
        shellSort(arr);
        System.out.println("Отсортированный массив: " + Arrays.toString(arr));
    }
}
